File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/direct/Java/Python/s910447421_commented.py
import sys
from math import *

ans = 100
balls = []
N = 0
taken = []

class FastScanner:
    def __init__(self):
        self.br = sys.stdin
        self.st = ""
    
    def next(self):
        while not self.st:
            try:
                self.st = self.br.readline().split()
            except:
                pass
        return self.st.pop(0)
    
    def nextInt(self):
        return int(self.next())
    
    def nextLong(self):
        return int(self.next())
    
    def nextDouble(self):
        return float(self.next())

class OutPut:
    def __init__(self):
        self.w = sys.stdout
    
    def pln(self, x):
        print(x, file=self.w)
        self.w.flush()
    
    def p(self, x):
        print(x, end='', file=self.w)
        self.w.flush()

def solve():
    global ans
    for add in range(1, N):
        taken = [False] * N
        p = balls[add][0] - balls[0][0]
        q = balls[add][1] - balls[0][1]
        ans = min(ans, check(p, q))

def check(p, q):
    ret = 0
    while True:
        added = False
        start = 0
        x = -1
        y = -1
        for i in range(N):
            if not taken[i]:
                added = True
                ret += 1
                start = i
                taken[i] = True
                x = balls[i][0] + p
                y = balls[i][1] + q
                break
        if not added:
            break
        for i in range(start + 1, N):
            if balls[i][0] == x and balls[i][1] == y:
                taken[i] = True
                x += p
                y += q
    return ret

def min(a, b):
    return math.min(a, b)

def max(a, b):
    return math.max(a, b)

def abs(x):
    return math.abs(x)

def ceil(num, den):
    ans = num // den
    if num % den != 0:
        ans += 1
    return ans

def GCD(a, b):
    if a == 0 or b == 0:
        return max(a, b)
    return GCD(min(a, b), max(a, b) % min(a, b))

def FastExp(base, exp, mod):
    ans = 1
    while exp > 0:
        if exp % 2 == 1:
            ans *= base
        exp //= 2
        base *= base
        base %= mod
        ans %= mod
    return ans

def ModInv(num, mod):
    return FastExp(num, mod - 2, mod)

def pop(x):
    cnt = 0
    while x > 0:
        if x % 2 == 1:
            cnt += 1
        x //= 2
    return cnt

I = FastScanner()
O = OutPut()
N = I.nextInt()
ans = N
balls = [[0, 0] for _ in range(N)]
for i in range(N):
    balls[i][0] = I.nextInt()
    balls[i][1] = I.nextInt()
balls.sort(key=lambda x: x[0])
solve()
for i in range(N):
    x = balls[i][0]
    y = balls[i][1]
    balls[i][0] = y
    balls[i][1] = x
balls.sort(key=lambda x: x[0])
solve()
O.pln(ans)  # End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/multiline/comment/direct/Java/C++/s910447421_commented.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int ans;
int N;
vector<pair<int, int>> balls;

void solve() {
    for (int add = 1; add < N; ++add) {
        int p = balls[add].first - balls[0].first;
        int q = balls[add].second - balls[0].second;
        vector<bool> taken(N, false);
        int current = check(p, q, taken);
        if (current < ans) ans = current;
    }
}

int check(int p, int q, vector<bool>& taken) {
    int ret = 0;
    while (true) {
        bool added = false;
        int start = 0;
        int x = -1, y = -1;
        for (int i = 0; i < N; ++i) {
            if (!taken[i]) {
                added = true;
                ret++;
                start = i;
                taken[i] = true;
                x = balls[i].first + p;
                y = balls[i].second + q;
                break;
            }
        }
        if (!added) break;
        for (int i = start + 1; i < N; ++i) {
            if (!taken[i] && balls[i].first == x && balls[i].second == y) {
                taken[i] = true;
                x += p;
                y += q;
            }
        }
    }
    return ret;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> N;
    ans = N;
    balls.resize(N);
    for (int i = 0; i < N; ++i) {
        cin >> balls[i].first >> balls[i].second;
    }
    sort(balls.begin(), balls.end());
    solve();
    for (auto& p : balls) swap(p.first, p.second);
    sort(balls.begin(), balls.end());
    solve();
    cout << ans << endl;
    return 0;
}
// End of Code
